File: Interview Preparation Kit/Dictionaries and Hashmaps/CountTriplets.cpp
// https://www.hackerrank.com/challenges/count-triplets-1/problem
#include <bits/stdc++.h>

using namespace std;

string ltrim(const string &);
string rtrim(const string &);
vector<string> split(const string &);

// Complete the countTriplets function below.
long countTriplets(vector<long> arr, long r) {
    map<long, long> first;
    map<long, long> second;
    long triplets = 0;

    for (int i = 0; i < arr.size(); i++) {
        long current = arr[i];
        if (current % r == 0) {
            long preceding = current / r;
            if (second.count(preceding) > 0) {
                triplets += second[preceding];
            }
            if (first.count(preceding) > 0) {
                if (second.count(current) > 0) {
                    second[current] += first[preceding];
                }
                else {
                    second[current] = first[preceding];
                }
            }
        }

        if (first.count(current) > 0) {
            first[current] += 1;
        }
        else {
            first[current] = 1;
        }
    }

    return triplets;
}

int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    string nr_temp;
    getline(cin, nr_temp);

    vector<string> nr = split(rtrim(nr_temp));

    int n = stoi(nr[0]);

    long r = stol(nr[1]);

    string arr_temp_temp;
    getline(cin, arr_temp_temp);

    vector<string> arr_temp = split(rtrim(arr_temp_temp));

    vector<long> arr(n);

    for (int i = 0; i < n; i++) {
        long arr_item = stol(arr_temp[i]);

        arr[i] = arr_item;
    }

    long ans = countTriplets(arr, r);

    fout << ans << "\n";

    fout.close();

    return 0;
}

string ltrim(const string &str) {
    string s(str);

    s.erase(
        s.begin(),
        find_if(s.begin(), s.end(), not1(ptr_fun<int, int>(isspace)))
    );

    return s;
}

string rtrim(const string &str) {
    string s(str);

    s.erase(
        find_if(s.rbegin(), s.rend(), not1(ptr_fun<int, int>(isspace))).base(),
        s.end()
    );

    return s;
}

vector<string> split(const string &str) {
    vector<string> tokens;

    string::size_type start = 0;
    string::size_type end = 0;

    while ((end = str.find(" ", start)) != string::npos) {
        tokens.push_back(str.substr(start, end - start));

        start = end + 1;
    }

    tokens.push_back(str.substr(start));

    return tokens;
}

File: Interview Preparation Kit/Dictionaries and Hashmaps/CountTriplets.java
// https://www.hackerrank.com/challenges/count-triplets-1/problem
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;
import java.util.regex.*;
import java.util.stream.*;
import static java.util.stream.Collectors.joining;
import static java.util.stream.Collectors.toList;

public class Solution {

    // Complete the countTriplets function below.
    static long countTriplets(List<Long> arr, long r) {
        HashMap<Long, Long> first = new HashMap<>();
        HashMap<Long, Long> second = new HashMap<>();
        long triplets = 0;

        for (int i = 0; i < arr.size(); i++) {
            long current = arr.get(i);
            if (current % r == 0) {
                long preceding = current / r;
                if (second.containsKey(preceding)) {
                    triplets += second.get(preceding);
                }
                if (first.containsKey(preceding)) {
                    if (second.containsKey(current)) {
                        second.put(current, second.get(current) + first.get(preceding));
                    }
                    else {
                        second.put(current, first.get(preceding));
                    }
                }
            }

            if (first.containsKey(current)) {
                first.put(current, first.get(current) + 1L);
            }
            else {
                first.put(current, 1L);
            }
        }

        return triplets;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        String[] nr = bufferedReader.readLine().replaceAll("\\s+$", "").split(" ");

        int n = Integer.parseInt(nr[0]);

        long r = Long.parseLong(nr[1]);

        List<Long> arr = Stream.of(bufferedReader.readLine().replaceAll("\\s+$", "").split(" "))
            .map(Long::parseLong)
            .collect(toList());

        long ans = countTriplets(arr, r);

        bufferedWriter.write(String.valueOf(ans));
        bufferedWriter.newLine();

        bufferedReader.close();
        bufferedWriter.close();
    }
}
